sym/src/convert.rs

```diff
@@ -1,312 +1,546 @@
-            .map(|x| x.0)
-            .map(Into::<SymbolicBitVec>::into)
-            .collect()
-    }
+            .map(|byte| byte.into_inner())
+            .flat_map(|bits| bits.into_iter()),
+    )
 }
 
-impl FromIterator<SymbolicBitVec> for SymbolicBitVec {
-    fn from_iter<T: IntoIterator<Item = SymbolicBitVec>>(iter: T) -> Self {
-        Self {
-            bits: iter
-                .into_iter()
-                .map(|bitvec| bitvec.bits.into_iter())
-                .flatten()
-                .collect(),
+pub fn concretize_bits_cow<'a, T, const N: usize>(
+    iter: impl Iterator<Item = std::borrow::Cow<'a, SymbolicBit>>,
+) -> std::result::Result<T, ConcretizationError>
+where
+    T: FromBytes<N>,
+{
+    // TODO Once we can use this directly in the function signature we can remove N
+    assert_eq!(std::mem::size_of::<T>(), N);
+
+    let mut bytes = [0u8; N];
+    let mut byte_index = 0;
+    let mut bit_index = 0;
+    for cow_bit in iter {
+        if byte_index >= N {
+            return Err(ConcretizationError::Overflow { max_bytes: N });
+        }
+
+        let bit = match cow_bit {
+            std::borrow::Cow::Owned(SymbolicBit::Literal(bit)) => bit,
+            std::borrow::Cow::Borrowed(SymbolicBit::Literal(bit)) => *bit,
+            _ => {
+                return Err(ConcretizationError::NonLiteralBit {
+                    bit_index,
+                    byte_index,
+                })
+            }
+        };
+
+        if bit {
+            bytes[byte_index] |= 1 << bit_index;
+        }
+
+        bit_index += 1;
+        if bit_index == 8 {
+            bit_index = 0;
+            byte_index += 1;
         }
     }
+
+    Ok(T::from_le_bytes(bytes))
 }
 
-fn into_symbitvec<T>(value: T) -> SymbolicBitVec
+pub fn concretize_bits<'a, T, const N: usize>(
+    iter: impl Iterator<Item = &'a SymbolicBit>,
+) -> std::result::Result<T, ConcretizationError>
 where
-    T: std::ops::BitAnd + std::ops::ShlAssign + PartialEq + From<u8> + Copy,
-    <T as std::ops::BitAnd>::Output: PartialEq + From<u8>,
+    T: FromBytes<N>,
 {
-    let num_bits = 8 * std::mem::size_of::<T>();
-    let mut bits = Vec::with_capacity(num_bits);
-    let mut mask: T = 1u8.into();
-    for _ in 0..num_bits {
-        bits.push(SymbolicBit::Literal(value & mask != 0u8.into()));
-        mask <<= 1u8.into();
-    }
+    concretize_bits_cow(iter.map(std::borrow::Cow::Borrowed))
+}
 
-    SymbolicBitVec { bits }
+pub fn concretize_bits_into<T, const N: usize>(
+    iter: impl IntoIterator<Item = SymbolicBit>,
+) -> std::result::Result<T, ConcretizationError>
+where
+    T: FromBytes<N>,
+{
+    concretize_bits_cow(iter.into_iter().map(std::borrow::Cow::Owned))
 }
 
-impl std::ops::Deref for SymbolicBitVec {
-    type Target = [SymbolicBit];
+/// Create an object from bytes. The object must be constructable from a known, fixed set of bytes.
+/// TODO Rename this trait since it also supports to bytes
+pub trait FromBytes<const N: usize> {
```

Comment:
This should be renamed and Rustdoc updated.. Perhaps `LittleEndianBytes` and removing the `le` abbreviation from the trait function names? Could conceivably generalize the `u8` to `T = u8` and naming `LittleEndian` instead so e.g. a bit-representation could still implement the trait.